packages/aragon-api/src/index.js

```diff
@@ -231,16 +274,49 @@ export class AppProxy {
         Promise.resolve(reducer(state, event))
       )
 
-    const store$ = initialState.pipe(
-      switchMap((initialState) =>
-        merge(
-          this.events(),
-          ...events
-        ).pipe(
-          mergeScan(wrappedReducer, initialState, 1),
-          map((state) => this.cache('state', state))
+    const store$ = combineLatest(cachedState$, initState$, cachedBlock$, latestBlock$).pipe(
+      switchMap(([cachedState, initState, cachedBlock, latestBlock]) => {
+        console.log('initState', initState)
+        console.log('cachedState', cachedState)
+        console.log(`cachedBlock ${cachedBlock} |  latestBlock: ${latestBlock}`)
+
+        // The block up to which to fetch past events.
+        // The reduced state up to this point will be cached on every load
+        const pastEventsToBlock = latestBlock - BLOCK_REORG_MARGIN
+
+        console.log(`--- pastEvents: ${cachedBlock} -> ${pastEventsToBlock}`)
+        console.log(`--- currentEvents$: from: ${pastEventsToBlock} -> future`)
+        // currentEvents$ is an observable for events starting from pastEventsToBlock
+        const currentEvents$ = this.events(pastEventsToBlock)
+
+        const pastState$ = this.pastEvents(cachedBlock, pastEventsToBlock).pipe(
+          // Prevent multiple subscriptions invoking duplicate calls
+          share(),
+          // single emission array of all pastEvents -> flatten to process events
+          flatMap(pastEvents => from(pastEvents)),
+          mergeScan(wrappedReducer, { ...cachedState, ...initState }, 1),
+          // debounce to reduce rendering and caching overthead
+          debounceTime(200),
+          tap((state) => {
+            console.count('cache pastState')
+            this.cache(CACHED_BLOCK_KEY, pastEventsToBlock)
+            this.cache('state', state)
+          })
+        )
+        const currentState$ = pastState$.pipe(
+          // Use the last past state as the initial state for reducing current/future states
+          last(),
+          switchMap(pastState => merge(currentEvents$, ...events).pipe(
```

Comment:
Something to be careful of here is to synchronize the external `events`'s cached block height with the app contract's cached block height (so we'd need to do the `pastEvents` for each of these external events).